x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/support/CachingUsernamePasswordRealm.java

```diff
@@ -78,74 +83,90 @@ public final void authenticate(AuthenticationToken authToken, ActionListener<Aut
+                        future.onResponse(new Tuple<>(result, userWithHash));
+                    } else {
+                        future.onResponse(new Tuple<>(result, null));
                     }
-                    listener.onResponse(result);
-                }, listener::onFailure));
+                }, future::onFailure));
             }
-        } else {
-            cache.invalidate(token.principal());
-            doAuthenticateAndCache(token, ActionListener.wrap((result) -> {
-                if (result.isAuthenticated()) {
-                    final User user = result.getUser();
-                    logger.debug("cached user came from a lookup and could not be used for authentication. " +
-                            "realm [{}] authenticated user [{}] with roles [{}]", name(), token.principal(), user.roles());
+
+            future.addListener(ActionListener.wrap(tuple -> {
+                if (tuple != null) {
+                    handleResult(future, createdAndStartedFuture.get(), token, tuple, listener);
+                } else {
+                    handleFailure(future, createdAndStartedFuture.get(), token, new IllegalStateException("unknown error authenticating"),
+                        listener);
                 }
-                listener.onResponse(result);
-            }, listener::onFailure));
+            }, e -> handleFailure(future, createdAndStartedFuture.get(), token, e, listener)),
+                threadPool.executor(ThreadPool.Names.GENERIC));
+        } catch (ExecutionException e) {
+            listener.onResponse(AuthenticationResult.unsuccessful("", e));
         }
     }
 
-    private void doAuthenticateAndCache(UsernamePasswordToken token, ActionListener<AuthenticationResult> listener) {
-        ActionListener<AuthenticationResult> wrapped = ActionListener.wrap((result) -> {
-            Objects.requireNonNull(result, "AuthenticationResult cannot be null");
-            if (result.getStatus() == AuthenticationResult.Status.SUCCESS) {
-                UserWithHash userWithHash = new UserWithHash(result.getUser(), token.credentials(), hasher);
-                // it doesn't matter if we already computed it elsewhere
-                cache.put(token.principal(), userWithHash);
+    private void handleResult(ListenableFuture<Tuple<AuthenticationResult, UserWithHash>> future, boolean createdAndStartedFuture,
+                              UsernamePasswordToken token, Tuple<AuthenticationResult, UserWithHash> result,
+                              ActionListener<AuthenticationResult> listener) {
+        final AuthenticationResult authResult = result.v1();
+        if (authResult == null) {
+            // this was from a lookup; clear and redo
+            cache.invalidate(token.principal(), future);
+            authenticateWithCache(token, listener);
+        } else if (authResult.isAuthenticated()) {
+            if (createdAndStartedFuture) {
```

Comment:
The use of `createdAndStartedFuture` here worries me.
We need to handle the two cases (directly authenticated vs loaded from cache) differently, but this variable is a couple of steps removed from that and I fear that something like a race condition somewhere in the cache could mean that this is `true`, but the user was not actually authenticated with the credentials from `token`.

Perhaps we could set an `AtomicReference` to `User` in the listener for `doAuthenticate` and check whether the `User` being pulled from the cache matches. Then were no longer testing "did we create the future" but "did we actually authenticate _this_ User object".

e.g.
```
 doAuthenticate(token, ActionListener.wrap(result -> {
                    if (result.isAuthenticated()) {
                        final User user = result.getUser();
                        authenticatedUser.set(user);
```
``` 
       final boolean wasAuthenticatedWithThisToken = tuple.v2() == authenticatedUser.get();
       handleResult(future, wasAuthenticatedWithThisToken, token, tuple, listener);
```